pkg/executor/join/base_semi_join.go

```diff
@@ -129,3 +132,58 @@ func (b *baseSemiJoin) concatenateProbeAndBuildRows(joinedChk *chunk.Chunk, sqlK
 	b.finishCurrentLookupLoop(joinedChk)
 	return nil
 }
+
+// Only used for semi and anti semi join
+func (b *baseSemiJoin) generateResultChkForRightBuildNoOtherCondition(resultChk *chunk.Chunk) {
+	if len(b.offsets) == 0 {
+		return
+	}
+
+	for index, colIndex := range b.lUsed {
+		srcCol := b.currentChunk.Column(colIndex)
+		dstCol := resultChk.Column(index)
+		chunk.CopyRows(dstCol, srcCol, b.offsets)
+	}
+
+	if len(b.lUsed) == 0 {
+		resultChk.SetNumVirtualRows(resultChk.NumRows() + len(b.offsets))
+	} else {
+		resultChk.SetNumVirtualRows(resultChk.NumRows())
+	}
+}
+
+// Only used for semi and anti semi join
+func (b *baseSemiJoin) generateResultChkForRightBuildWithOtherCondition(remainCap int, chk *chunk.Chunk, resultRows []bool) {
```

Comment:
how about add another  parameter `expectedResult` to this function, and only generate the result if `resultRows[i] == expectedResult`? In this way, `isMatchedRowsAdjustedForOutput` can be removed and there is no need to adjust `b.isMatchedRows` before generate result for anti semi join

Reply:
> how about add another parameter `expectedResult` to this function, and only generate the result if `resultRows[i] == expectedResult`? In this way, `isMatchedRowsAdjustedForOutput` can be removed and there is no need to adjust `b.isMatchedRows` before generate result for anti semi join

done